src/isolate.rs

```diff
@@ -864,24 +884,227 @@ impl IsolateHandle {
+    (*cxx_isolate).exit();
+    ScopeData::get_root_mut(&mut *cxx_isolate);
+    // construct V8's native Unlocker instance
+    v8__Unlocker__CONSTRUCT(&mut locker, cxx_isolate);
+    locker
+  }
+}
+
+impl<'a> Drop for Unlocker<'a> {
+  fn drop(&mut self) {
+    unsafe {
+      v8__Unlocker__DESTRUCT(self);
+      // restore the entered isolate
+      (*self.cxx_isolate).enter();
+    }
+  }
+}
+
+/// An entered isolate within a Locker's scope. Created when a SharedIsolate is borrowed.
+/// When dropped, the isolate is exited and the locker is deconstructed.
+/// This is similar to V8's Isolate::Scope, but a Locker is implicitly entered for the life of a LockedIsolate.
+pub struct LockedIsolate<'a> {
+  cxx_isolate: &'a mut NonNull<Isolate>,
+  locker: Locker,
+}
+
+impl<'a> LockedIsolate<'a> {
+  /// Creates a new Locker scope and enters the isolate within, returning a scoped LockedIsolate.
+  pub(crate) fn new(cxx_isolate: &'a mut NonNull<Isolate>) -> Self {
+    let locker = unsafe {
+      let locker = Locker::new(cxx_isolate.as_ptr());
+      cxx_isolate.as_mut().enter();
+      locker
+    };
+    Self {
+      cxx_isolate,
+      locker,
+    }
+  }
+
+  /// Borrows a new Unlocker, temporarily releasing V8's lock on the current thread.
+  /// The Unlocker exits the LockedIsolate and signals to V8 that the isolates locked to this thread may be used elsewhere.
+  /// When the guard is dropped, the underlying Unlocker is destroyed and the isolate is re-entered.
+  ///
+  /// This function is marked unsafe because Rust's rules allow you to interact with other LockedIsolates while an Unlocker is active, which is not allowed.
+  /// To make up for this limitation, the embedder should instead drop all active LockedIsolate references before doing long-lasting blocking work, re-entering afterwards.
+  pub unsafe fn unlock(&mut self) -> Unlocker<'_> {
+    Unlocker::new(self.cxx_isolate.as_mut())
+  }
+}
+
+impl<'a> Deref for LockedIsolate<'a> {
+  type Target = Isolate;
+  fn deref(&self) -> &Self::Target {
+    unsafe { self.cxx_isolate.as_ref() }
+  }
+}
+
+impl<'a> DerefMut for LockedIsolate<'a> {
+  fn deref_mut(&mut self) -> &mut Self::Target {
+    unsafe { self.cxx_isolate.as_mut() }
+  }
+}
+
+impl<'a> AsMut<Isolate> for LockedIsolate<'a> {
+  fn as_mut(&mut self) -> &mut Isolate {
+    self
+  }
+}
+
+impl<'a> Drop for LockedIsolate<'a> {
+  fn drop(&mut self) {
+    // When a Locker is dropped, v8 automatically exits any active scopes.
+    // Resetting to the root scope exits any zombie scopes so that the next LockedIsolate can cleanly re-enter a context.
+    ScopeData::get_root_mut(self);
+    unsafe { self.exit() }
+  }
+}
+
+/// A reference to an Isolate that may be sent between threads and borrowed with a Locker before entering V8.
+/// May be created by calling Isolate::new_shared(params), or converted via SharedIsolate::from(OwnedIsolate).
+#[derive(Debug)]
+pub struct SharedIsolate {
```

Comment:
Ideally the `IsolateHandle` type (returned from `Isolate::thread_safe_handle()`) would be used for this.

Reply:
I played around with this idea originally. Because `IsolateHandle` is `Clone` we'd need to introduce a mutex around the isolate, and I tried to veer away from adding any locks (maybe the embedded would like to bring their own lock). Also, an `OwnedIsolate` is entered for its entire lifespan so we'd need a way to break out of that locker without dropping the isolate, which would probably need to introduce a breaking change.

It's worth noting that `SharedIsolate` impl's the same deref as the `IsolateHandle`, so the related methods can be accessed. I'd love to hear out and implement any ideas on this front.

Reply:
My mistake, `SharedIsolate` was actually implementing deref for `Isolate`, which was no good. I've corrected it by making `SharedIsolate` deref as `IsolateHandle`, but that still doesn't solve the main issue here.

Reply:
I'm really having trouble now keeping all the different Isolate types in my head and understanding the differences between them. I wonder if we could do something radically more simple where:
* Using lockers is no longer optional, but required.
* Isolates are reference counted. `Isolate` clones and `Global<_>` handles both establish a reference to the v8 isolate. The isolate is automatically disposed when the reference count reaches 0. Note that we are already kind of keeping a reference count in `IsolateAnnex`.

![image](https://user-images.githubusercontent.com/218257/127556366-709cbe65-89e4-46d5-83f6-22190812ad44.png)

Reply:
I've started to converge toward what you've described in the latest commit. In this case, I've kept `IsolateHandle` and `Isolate` similar to main branch to avoid having to shuffle all the existing methods around. OwnedIsolate, LockedIsolate, SharedIsolate, etc. are now gone.

- `Isolate::new_handle()` returns an IsolateHandle that can be cloned/shared and locked. 
- `Isolate::new()` sort-of maintains the existing behavior of returning a `Locker` that will also dispose of the Isolate when dropped. We can ditch this idea and force the shared IsolateHandle path, but that adds a new line to every test case and is a wider breaking change.

The IsolateHandle is used to guard the Isolate reference count and dispose of it when dropped.

`Locker` can be obtained from an `IsolateHandle`, which derefs `Isolate` and can be used to create `HandleScope`. 
It has a zero-sized type parameter to allow the creator (`Isolate::new()` or `IsolateHandle::lock(&mut self)`) to create a new 'borrowed' Locker *or*  use an owned one (default), which lets us make the lifetime on Locker trivially optional.

It was relatively straightforward to clean up SnapshotCreator's `get_isolate()` too with this system.